lib/src/io_client.dart

```diff
@@ -23,42 +24,82 @@ class IOClient extends BaseClient {
 
   /// Sends an HTTP request and asynchronously returns the response.
   @override
-  Future<IOStreamedResponse> send(BaseRequest request) async {
+  Future<IOStreamedResponse> send(BaseRequest request, {Duration? timeout}) {
+    final completer = Completer<IOStreamedResponse>();
+    _send(request, timeout, completer);
+    return completer.future;
+  }
+
+  Future<void> _send(BaseRequest request, Duration? timeout,
+      Completer<IOStreamedResponse> completer) async {
     var stream = request.finalize();
 
+    Timer? timer;
+    late void Function() onTimeout;
```

Comment:
Can make this nullable instead of late, and do `onTimeout?.call();` or `onTimeout!()` in the timer callback.
That could potentially be cheaper than a late field. (Unless the compiler is being competent about late non-nullable fields and using `null` as placeholder anyway)

Reply:
I think I have a slight preference for `late` as it matches the intended semantics more closely IMO. Outside of performance concerns do you have a preference one way or the other?

Reply:
I consider `late` an unsafe and dynamic feature at the same level of as, well, `dynamic`. I try to avoid it as much as possible, just as I do `dynamic`. Late variables have their uses, where it makes things easier, but it never makes them *safer*. 

Using `late` (and `dynamic`) allows statically undetectable run-time errors with visible marker at the use position. I prefer to use something nullable and `!` when possible *because* it is more verbose and therefore highlights the assumption at the read (by adding `!`).

This is a fairly fundamental library, and I think we should always choose safety (and performance) over brevity.

Reply:
Using `late` here I believe is more about choosing something that matches the underlying semantics/implementation than brevity. I don't feel super strongly either way, but this is always assigned, and I don't agree it is any safer to use `!` everywhere - we would not be adding any null checks either way.

Reply:
Hmm actually on re-reading it I think this is only assigned if `timeout` is non-null, in which case making it nullable does match the semantics better, so I retract my previous comment and agree it should be nullable.

Reply:
I wasn't taking enough advantage of flow analysis. Turns out we don't need `late` or `?`. If I switch a few lines around then definite assignment analysis takes over.